File: scripts/upstream_authority_registry.py
"""Shared fail-closed reader for upstream contract authority pins."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any
from urllib.parse import urlparse


REGISTRY_PATH = Path(__file__).resolve().parent.parent / "benchmarks" / "authorities.json"
REQUIRED_AUTHORITIES = {
    "dspy_stable",
    "dspy_instruction_optimizers",
    "gepa_v0_1_1_contract",
    "gepa_v0_1_4_contract",
    "optimize_anything_artifact",
    "req_llm",
    "swe_bench_verified",
}
REQUIRED_CONTRACTS = {
    "dspy_stable_upstream_fidelity": "dspy_stable",
    "optimize_anything_swe_bench_flask_5014_dataset": "swe_bench_verified",
    "optimize_anything_upstream_differential_protocol": "optimize_anything_artifact",
    "req_llm_beam_runtime_dependency": "req_llm",
    "t1_gepa_v014_structural_differential_contract": "gepa_v0_1_4_contract",
    "t1_gepa_v011_structural_differential_contract": "gepa_v0_1_1_contract",
    "t1_instruction_optimizer_differential_contract": "dspy_instruction_optimizers",
}


def load_registry_authority(
    contract_id: str, path: Path = REGISTRY_PATH
) -> tuple[dict[str, Any], dict[str, Any]]:
    try:
        ledger = json.loads(path.read_text())
        if ledger.get("schema_version") != 1:
            raise ValueError("unsupported schema_version")
        pinned_sources = ledger["pinned_sources"]
        contracts = ledger["contracts"]
        if not REQUIRED_AUTHORITIES.issubset(pinned_sources):
            raise ValueError("missing required authorities")
        if not REQUIRED_CONTRACTS.keys() <= contracts.keys():
            raise ValueError("missing required contracts")
        for required_contract, expected_authority in REQUIRED_CONTRACTS.items():
            actual_authority = contracts[required_contract].get("authority")
            if actual_authority != expected_authority:
                raise ValueError(
                    f"contract {required_contract} must bind authority "
                    f"{expected_authority}, got {actual_authority!r}"
                )

        contract_ids_by_authority: dict[str, list[str]] = {}
        for candidate_id, candidate in contracts.items():
            authority_id = candidate["authority"]
            contract_ids_by_authority.setdefault(authority_id, []).append(candidate_id)

        authorities = {
            authority_id: _project_authority(
                authority_id,
                pinned_sources[authority_id],
                contract_ids,
            )
            for authority_id, contract_ids in contract_ids_by_authority.items()
        }

        for candidate_id, candidate in authorities.items():
            required = (
                "project",
                "repository",
                "version",
                "git_ref",
                "commit",
                "source_hashes",
                "contract_ids",
            )
            if any(not candidate.get(key) for key in required):
                raise ValueError(f"authority {candidate_id} is missing required fields")
            if not _digest(candidate["commit"], 40):
                raise ValueError(f"authority {candidate_id} has an invalid commit")
            if not candidate["source_hashes"] or any(
                not _digest(digest, 64) for digest in candidate["source_hashes"].values()
            ):
                raise ValueError(f"authority {candidate_id} has invalid source hashes")
        for candidate_id, candidate in contracts.items():
            candidate_authority = authorities[candidate["authority"]]
            if candidate_id not in candidate_authority["contract_ids"]:
                raise ValueError(
                    f"contract {candidate_id} is incompatible with authority {candidate['authority']}"
                )
        declared_contracts = [
            candidate_id
            for candidate in authorities.values()
            for candidate_id in candidate["contract_ids"]
        ]
        if sorted(declared_contracts) != sorted(contracts):
            raise ValueError("contract identifiers must be declared exactly once")
        authority = authorities[contracts[contract_id]["authority"]]
        registry = {
            "schema_version": 1,
            "authorities": authorities,
            "contracts": contracts,
        }
        return registry, authority
    except (OSError, KeyError, TypeError, json.JSONDecodeError, ValueError) as exc:
        raise RuntimeError(f"invalid upstream authority registry {path.resolve()}: {exc}") from exc


def _project_authority(
    authority_id: str,
    source: dict[str, Any],
    contract_ids: list[str],
) -> dict[str, Any]:
    authority = dict(source)
    repository = authority["repository"]
    project = urlparse(repository).path.strip("/")
    if not project:
        raise ValueError(f"authority {authority_id} has an invalid repository URL")

    source_hashes = authority.get("source_hashes")
    if not source_hashes:
        files = authority.get("files")
        if not files:
            raise ValueError(f"authority {authority_id} has no exact source hashes")
        source_hashes = {entry["path"]: entry["sha256"] for entry in files}

    authority["project"] = project
    authority["source_hashes"] = source_hashes
    authority["contract_ids"] = sorted(contract_ids)
    return authority


def source_hash_failures(actual_hashes: dict[str, str], authority: dict[str, Any]) -> list[str]:
    expected_hashes = authority["source_hashes"]
    failures = []
    if set(actual_hashes) != set(expected_hashes):
        failures.append(f"source set: expected {sorted(expected_hashes)}, got {sorted(actual_hashes)}")
    for relative_path in sorted(set(actual_hashes) & set(expected_hashes)):
        actual_hash = actual_hashes[relative_path]
        expected_hash = expected_hashes[relative_path]
        if actual_hash != expected_hash:
            failures.append(
                f"{relative_path}: expected sha256 {expected_hash}, got {actual_hash}"
            )
    return failures


def _digest(value: Any, length: int) -> bool:
    return (
        isinstance(value, str)
        and len(value) == length
        and all(char in "0123456789abcdef" for char in value)
    )

File: bench/imp/upstream_authority_registry.ex
defmodule Imp.UpstreamAuthorityRegistry do
  @moduledoc false

  @registry_path Path.expand("../../benchmarks/authorities.json", __DIR__)
  @required_authorities ~w(
    dspy_stable
    dspy_instruction_optimizers
    gepa_v0_1_1_contract
    optimize_anything_artifact
    req_llm
    swe_bench_verified
  )
  @required_contracts %{
    "dspy_stable_upstream_fidelity" => "dspy_stable",
    "optimize_anything_swe_bench_flask_5014_dataset" => "swe_bench_verified",
    "optimize_anything_upstream_differential_protocol" => "optimize_anything_artifact",
    "req_llm_beam_runtime_dependency" => "req_llm",
    "t1_gepa_v011_structural_differential_contract" => "gepa_v0_1_1_contract",
    "t1_instruction_optimizer_differential_contract" => "dspy_instruction_optimizers"
  }
  @hex40 ~r/\A[0-9a-f]{40}\z/
  @hex64 ~r/\A[0-9a-f]{64}\z/

  @doc false
  def path, do: @registry_path

  @doc false
  def load!(path \\ @registry_path) do
    ledger =
      path
      |> File.read!()
      |> Jason.decode!()

    ledger
    |> project!(path)
    |> validate!(path)
  rescue
    error in [File.Error, Jason.DecodeError, ArgumentError, KeyError] ->
      reraise ArgumentError,
              [
                message:
                  "invalid upstream authority ledger #{Path.expand(path)}: #{Exception.message(error)}"
              ],
              __STACKTRACE__
  end

  @doc false
  def authority!(registry, contract_id) do
    contract = registry |> Map.fetch!("contracts") |> Map.fetch!(contract_id)
    authority_id = Map.fetch!(contract, "authority")
    registry |> Map.fetch!("authorities") |> Map.fetch!(authority_id)
  rescue
    KeyError ->
      reraise ArgumentError,
              [message: "upstream authority registry has no contract #{inspect(contract_id)}"],
              __STACKTRACE__
  end

  defp project!(
         %{
           "schema_version" => 1,
           "pinned_sources" => pinned_sources,
           "contracts" => contracts
         },
         path
       )
       when is_map(pinned_sources) and is_map(contracts) do
    validate_required_bindings!(contracts)

    contract_ids_by_authority =
      Enum.reduce(contracts, %{}, fn
        {contract_id, %{"authority" => authority_id}}, acc
        when is_binary(contract_id) and is_binary(authority_id) ->
          Map.update(acc, authority_id, [contract_id], &[contract_id | &1])

        _entry, _acc ->
          raise ArgumentError, "contracts in #{Path.expand(path)} must name one authority"
      end)

    authorities =
      Map.new(contract_ids_by_authority, fn {authority_id, contract_ids} ->
        source =
          Map.get(pinned_sources, authority_id) ||
            raise(ArgumentError, "contract authority #{authority_id} is not pinned")

        authority =
          source
          |> Map.put("project", project_name!(source, authority_id))
          |> Map.put("source_hashes", source_hashes!(source, authority_id))
          |> Map.put("contract_ids", Enum.sort(contract_ids))

        {authority_id, authority}
      end)

    %{
      "schema_version" => 1,
      "authorities" => authorities,
      "contracts" => contracts
    }
  end

  defp project!(_ledger, _path) do
    raise ArgumentError,
          "expected schema_version 1 with pinned_sources and canonical contract bindings"
  end

  defp source_hashes!(%{"source_hashes" => hashes}, _id)
       when is_map(hashes) and map_size(hashes) > 0,
       do: hashes

  defp source_hashes!(%{"files" => files}, id) when is_list(files) and files != [] do
    Map.new(files, fn
      %{"path" => path, "sha256" => sha256}
      when is_binary(path) and is_binary(sha256) ->
        {path, sha256}

      _file ->
        raise ArgumentError, "authority #{id} has an invalid pinned file"
    end)
  end

  defp source_hashes!(_source, id),
    do: raise(ArgumentError, "authority #{id} has no exact source hashes")

  defp project_name!(%{"repository" => repository}, id) when is_binary(repository) do
    case URI.parse(repository) do
      %URI{host: host, path: path} when is_binary(host) and is_binary(path) and path != "" ->
        String.trim(path, "/")

      _uri ->
        raise ArgumentError, "authority #{id} has an invalid repository URL"
    end
  end

  defp project_name!(_source, id),
    do: raise(ArgumentError, "authority #{id} has no repository URL")

  defp validate!(
         %{"schema_version" => 1, "authorities" => authorities, "contracts" => contracts} =
           registry,
         path
       )
       when is_map(authorities) and is_map(contracts) and map_size(authorities) > 0 and
              map_size(contracts) > 0 do
    require_names!(authorities, @required_authorities, "authorities")
    require_names!(contracts, Map.keys(@required_contracts), "contracts")
    validate_required_bindings!(contracts)
    Enum.each(authorities, &validate_authority!/1)
    Enum.each(contracts, &validate_contract!(&1, authorities))

    declared =
      authorities
      |> Enum.flat_map(fn {_id, authority} -> authority["contract_ids"] end)
      |> Enum.sort()

    actual = contracts |> Map.keys() |> Enum.sort()

    if declared != actual do
      raise ArgumentError,
            "contract identifiers must be declared exactly once by their authorities in #{Path.expand(path)}"
    end

    registry
  end

  defp validate!(_registry, _path),
    do: raise(ArgumentError, "expected schema_version 1 with non-empty authorities and contracts")

  defp validate_authority!({id, authority}) when is_binary(id) and is_map(authority) do
    Enum.each(["project", "repository", "version", "git_ref", "commit"], fn key ->
      require_string!(authority, key, "authority #{id}")
    end)

    require_hash!(authority["commit"], @hex40, "authority #{id} commit")

    hashes = Map.fetch!(authority, "source_hashes")

    unless is_map(hashes) and map_size(hashes) > 0 do
      raise ArgumentError, "authority #{id} source_hashes must be a non-empty object"
    end

    Enum.each(hashes, fn {name, hash} ->
      require_hash!(hash, @hex64, "authority #{id} source hash #{inspect(name)}")
    end)

    contract_ids = Map.fetch!(authority, "contract_ids")

    unless is_list(contract_ids) and contract_ids != [] and
             Enum.all?(contract_ids, &(is_binary(&1) and &1 != "")) and
             length(contract_ids) == length(Enum.uniq(contract_ids)) do
      raise ArgumentError, "authority #{id} contract_ids must be a non-empty unique string list"
    end
  end

  defp validate_authority!(_), do: raise(ArgumentError, "authority entries must be named objects")

  defp validate_contract!({contract_id, %{"authority" => authority_id}}, authorities)
       when is_binary(contract_id) and contract_id != "" and is_binary(authority_id) do
    authority =
      Map.get(authorities, authority_id) ||
        raise(ArgumentError, "contract #{contract_id} names missing authority #{authority_id}")

    unless contract_id in authority["contract_ids"] do
      raise ArgumentError,
            "contract #{contract_id} is not declared by authority #{authority_id}"
    end
  end

  defp validate_contract!(_, _),
    do: raise(ArgumentError, "contract entries must name one authority")

  defp require_names!(entries, required, kind) do
    missing = Enum.reject(required, &Map.has_key?(entries, &1))

    if missing != [],
      do: raise(ArgumentError, "missing required #{kind}: #{Enum.join(missing, ", ")}")
  end

  defp validate_required_bindings!(contracts) do
    Enum.each(@required_contracts, fn {contract_id, expected_authority} ->
      actual_authority = get_in(contracts, [contract_id, "authority"])

      unless actual_authority == expected_authority do
        raise ArgumentError,
              "contract #{contract_id} must bind authority #{expected_authority}, got #{inspect(actual_authority)}"
      end
    end)
  end

  defp require_string!(map, key, context) do
    case Map.fetch!(map, key) do
      value when is_binary(value) and value != "" -> value
      _ -> raise ArgumentError, "#{context} #{key} must be a non-empty string"
    end
  end

  defp require_hash!(value, pattern, context) do
    unless is_binary(value) and Regex.match?(pattern, value) do
      raise ArgumentError, "#{context} must be a lowercase SHA digest"
    end
  end
end
